### ISSUE TITLE
For_each Key Issue

### ISSUE THREAD

**Issue by tikicoder** [hashicorp/terraform#24448]
### Terraform Version

Terraform v0.12.23
provider.azurev2.2.0

### Terraform Configuration Files
```
FILE multiRegion.tfvars

regions = {
  northcentralus = {
    location = "northcentralus"
  }
  southcentralus = {
    location = "southcentralus"
  }
}
```
```
FILE appservice/appservice.tf

bad
keys = {
  for item in keys(var.regions):
    item => item
}
  
good
keys = {
  for item in keys(var.regions):
    "_${item}" => item
}
```
### Debug Output

N/A

### Crash Output

N/A

### Expected Behavior

I should be able to use keys that are location-based without having to add _ to them.  

### Actual Behavior

When you run the fail version it is unable to create the azurerm_app_service because it cannout find the key.

### Steps to Reproduce

./terraform plan --var-file=multiRegion.tfvars

### References

I have attached a working and not working versions

[terraformTestFail.zip](https://github.com/hashicorp/terraform/files/4378520/terraformTestFail.zip)
[terraformTestWorks.zip](https://github.com/hashicorp/terraform/files/4378521/terraformTestWorks.zip)

**Comment by jbardin** [comment]
Hi @tikicoder,

Can you help me understand what the failure case is here in the for expressions? The "bad" example seems to produce what I would expect, which is that the output matches the "good" example with the keys prefixed by `_`.

The output from the above code is:

```
bad = {
  "northcentralus" = "northcentralus"
  "southcentralus" = "southcentralus"
}
good = {
  "_northcentralus" = "northcentralus"
  "_southcentralus" = "southcentralus"
}
```

**Comment by tikicoder** [comment]
I hate to ask but did you create the resources?  The bad doesn't create the resources and fails on the line (61):

app_service_plan_id = azurerm_app_service_plan.main[each.value.key].id

It cannot find the key northcentralus.  If I use keys that are _northcentralus or northcentralus_ it works.

**Comment by jbardin** [comment]
@tikicoder,

Could you please add the actual error output?

Creating resource with these interpolations works correctly, at least in the basic case. Since the `service` and `service_plan` are using different `for_each` maps, It could be possible that there are keys that don't correlate between the two though I'm not sure how from this example. 

If it is the line you mention, the issue should simply be that there are `"key"` values in `local_topic_region` that are not in `local.keys`, but we need to see what the actual error to be certain.

**Comment by tikicoder** [comment]
Here is an updated file, and the exact error.
Error: Invalid index

  on appservice/appservice.tf line 61, in resource "azurerm_app_service" "main":
  61:   app_service_plan_id = azurerm_app_service_plan.main[each.key].id
    |----------------
    | azurerm_app_service_plan.main is object with no attributes
    | each.key is "northcentralus"

The given key does not identify an element in this collection value.


Error: Invalid index

  on appservice/appservice.tf line 61, in resource "azurerm_app_service" "main":
  61:   app_service_plan_id = azurerm_app_service_plan.main[each.key].id
    |----------------
    | azurerm_app_service_plan.main is object with no attributes
    | each.key is "southcentralus"

The given key does not identify an element in this collection value.
[terraformTestFail.zip](https://github.com/hashicorp/terraform/files/4382665/terraformTestFail.zip)

**Comment by jbardin** [comment]
Thanks @tikicoder,

The `azurerm_app_service_plan.main is object with no attributes` is the part I needed.

This looks like the same issue discussed in #21170.
Can you trigger this from a clean working directory, or is there prior state that is required?

It may help to remove the dependency from azurerm in your example as well, as is should have no bearing on the mechanics of resources using `for_each`.

**Comment by tikicoder** [comment]
From a clean working directory is fine.  That is typically how I have ran it.

The issue it only fails on the resources, everything else the key works just fine.  The ticket you mentioned and this is definitely related.  I have never tried to use numeric keys I have not had an issue accessing things like that ticket using my key.  Right now I have a TF setup that is using foreach to create multiple resources via modules and then grabbing the values back and using them in other modules.  However, again my keys are strings and not numeric, it seems to work better that way, I have noticed there is a toset but I haven't used it.

**Comment by tikicoder** [comment]
I did just run into the same issue as #21170.  So referencing the key via the module.  I did a nasty workaround but I have something that should work.

So it looks like the issues are similar but still probably different.

Here is what I ended up having to do to get it to work.
length(keys(module.networking.subnet_asp_name)) < 1 ? "" : module.networking.subnet_asp_name[keys(module.networking.subnet_asp_name)[index(keys(var.settings_networking), item)]][0],

Basically look up the index of the key based off the settings variable I sent.

Here is some extra information.

I was doing some debugging. 
keys(module.networking.subnet_function_name)[0] == item true
index(keys(module.networking.subnet_function_name)[0], item) error could not find item
module.networking.subnet_function_name[item] item was not a valid option.

**Comment by tikicoder** [comment]
Someone tried this and it looks like it might work

module.networking.subnet_function_name[format("_%s", item)]

it's almost like some aspects it's able to be treated like a string and in others its not.

**Comment by danieldreier** [comment]
@tikicoder I have copied your failure test case into a GitHub repo at https://github.com/danieldreier/terraform-issue-reproductions/tree/master/24448 so that we can iterate on it together. Here's the output I'm seeing running it:

```
bash-3.2$ terraform-v0.12.23 init
Initializing modules...

Initializing the backend...

Initializing provider plugins...

The following providers do not have any version constraints in configuration,
so the latest version was installed.

To prevent automatic upgrades to new major versions that may contain breaking
changes, it is recommended to add version = "..." constraints to the
corresponding provider blocks in configuration, with the constraint strings
suggested below.

* provider.azurerm: version = "~> 2.14"

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
bash-3.2$ terraform-v0.12.23 plan --var-file=multiRegion.tfvars
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.


------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # module.appservice.azurerm_app_service.main["northcentralus"] will be created
  + resource "azurerm_app_service" "main" {
      + app_service_plan_id            = (known after apply)
      + app_settings                   = (known after apply)
      + client_affinity_enabled        = (known after apply)
      + default_site_hostname          = (known after apply)
      + enabled                        = true
      + https_only                     = false
      + id                             = (known after apply)
      + location                       = "northcentralus"
      + name                           = "north-northcentralus-appservice"
      + outbound_ip_addresses          = (known after apply)
      + possible_outbound_ip_addresses = (known after apply)
      + resource_group_name            = "rontest-bold-resources"
      + site_credential                = (known after apply)
      + source_control                 = (known after apply)

      + auth_settings {
          + additional_login_params        = (known after apply)
          + allowed_external_redirect_urls = (known after apply)
          + default_provider               = (known after apply)
          + enabled                        = (known after apply)
          + issuer                         = (known after apply)
          + runtime_version                = (known after apply)
          + token_refresh_extension_hours  = (known after apply)
          + token_store_enabled            = (known after apply)
          + unauthenticated_client_action  = (known after apply)

          + active_directory {
              + allowed_audiences = (known after apply)
              + client_id         = (known after apply)
              + client_secret     = (sensitive value)
            }

          + facebook {
              + app_id       = (known after apply)
              + app_secret   = (sensitive value)
              + oauth_scopes = (known after apply)
            }

          + google {
              + client_id     = (known after apply)
              + client_secret = (sensitive value)
              + oauth_scopes  = (known after apply)
            }

          + microsoft {
              + client_id     = (known after apply)
              + client_secret = (sensitive value)
              + oauth_scopes  = (known after apply)
            }                                                                                                                                      [188/1980]

          + twitter {
              + consumer_key    = (known after apply)
              + consumer_secret = (sensitive value)
            }
        }

      + connection_string {
          + name  = (known after apply)
          + type  = (known after apply)
          + value = (sensitive value)
        }

      + identity {
          + identity_ids = (known after apply)
          + principal_id = (known after apply)
          + tenant_id    = (known after apply)
          + type         = (known after apply)
        }

      + logs {
          + application_logs {
              + azure_blob_storage {
                  + level             = (known after apply)
                  + retention_in_days = (known after apply)
                  + sas_url           = (sensitive value)
                }
            }

          + http_logs {
              + azure_blob_storage {
                  + retention_in_days = (known after apply)
                  + sas_url           = (sensitive value)
                }

              + file_system {
                  + retention_in_days = (known after apply)
                  + retention_in_mb   = (known after apply)
                }
            }
        }

      + site_config {
          + always_on                = false
          + dotnet_framework_version = "v4.0"
          + ftps_state               = (known after apply)
          + http2_enabled            = false
          + ip_restriction           = (known after apply)
          + linux_fx_version         = (known after apply)
          + local_mysql_enabled      = (known after apply)
          + managed_pipeline_mode    = (known after apply)
          + min_tls_version          = (known after apply)
          + remote_debugging_enabled = false
          + remote_debugging_version = (known after apply)
          + scm_type                 = "None"
          + websockets_enabled       = (known after apply)
          + windows_fx_version       = (known after apply)

          + cors {
              + allowed_origins     = (known after apply)
              + support_credentials = (known after apply)
            }
        }

      + storage_account {
          + access_key   = (sensitive value)
          + account_name = (known after apply)
          + mount_path   = (known after apply)
          + name         = (known after apply)
          + share_name   = (known after apply)
          + type         = (known after apply)
        }
    }

  # module.appservice.azurerm_app_service.main["southcentralus"] will be created
  + resource "azurerm_app_service" "main" {
      + app_service_plan_id            = (known after apply)
      + app_settings                   = (known after apply)
      + client_affinity_enabled        = (known after apply)
      + default_site_hostname          = (known after apply)
      + enabled                        = true
      + https_only                     = false
      + id                             = (known after apply)
      + location                       = "southcentralus"
      + name                           = "south-southcentralus-appservice"
      + outbound_ip_addresses          = (known after apply)
      + possible_outbound_ip_addresses = (known after apply)
      + resource_group_name            = "rontest-bold-resources"
      + site_credential                = (known after apply)
      + source_control                 = (known after apply)

      + auth_settings {
          + additional_login_params        = (known after apply)
          + allowed_external_redirect_urls = (known after apply)
          + default_provider               = (known after apply)
          + enabled                        = (known after apply)
          + issuer                         = (known after apply)
          + runtime_version                = (known after apply)
          + token_refresh_extension_hours  = (known after apply)
          + token_store_enabled            = (known after apply)
          + unauthenticated_client_action  = (known after apply)

          + active_directory {
              + allowed_audiences = (known after apply)
              + client_id         = (known after apply)
              + client_secret     = (sensitive value)
            }

          + facebook {
              + app_id       = (known after apply)
              + app_secret   = (sensitive value)
              + oauth_scopes = (known after apply)
            }

          + google {
              + client_id     = (known after apply)
              + client_secret = (sensitive value)
              + oauth_scopes  = (known after apply)
            }

          + microsoft {
              + client_id     = (known after apply)
              + client_secret = (sensitive value)
              + oauth_scopes  = (known after apply)
            }
                                                                                                                                                    [62/1980]
          + twitter {
              + consumer_key    = (known after apply)
              + consumer_secret = (sensitive value)
            }
        }

      + connection_string {
          + name  = (known after apply)
          + type  = (known after apply)
          + value = (sensitive value)
        }

      + identity {
          + identity_ids = (known after apply)
          + principal_id = (known after apply)
          + tenant_id    = (known after apply)
          + type         = (known after apply)
        }

      + logs {
          + application_logs {
              + azure_blob_storage {
                  + level             = (known after apply)
                  + retention_in_days = (known after apply)
                  + sas_url           = (sensitive value)
                }
            }

          + http_logs {
              + azure_blob_storage {
                  + retention_in_days = (known after apply)
                  + sas_url           = (sensitive value)
                }

              + file_system {
                  + retention_in_days = (known after apply)
                  + retention_in_mb   = (known after apply)
                }
            }
        }

      + site_config {
          + always_on                = false
          + dotnet_framework_version = "v4.0"
          + ftps_state               = (known after apply)
          + http2_enabled            = false
          + ip_restriction           = (known after apply)
          + linux_fx_version         = (known after apply)
          + local_mysql_enabled      = (known after apply)
          + managed_pipeline_mode    = (known after apply)
          + min_tls_version          = (known after apply)
          + remote_debugging_enabled = false
          + remote_debugging_version = (known after apply)
          + scm_type                 = "None"
          + websockets_enabled       = (known after apply)
          + windows_fx_version       = (known after apply)

          + cors {
              + allowed_origins     = (known after apply)
              + support_credentials = (known after apply)
            }
        }

      + storage_account {
          + access_key   = (sensitive value)
          + account_name = (known after apply)
          + mount_path   = (known after apply)
          + name         = (known after apply)
          + share_name   = (known after apply)
          + type         = (known after apply)
        }
    }

  # module.appservice.azurerm_app_service_plan.main["northcentralus"] will be created
  + resource "azurerm_app_service_plan" "main" {
      + id                           = (known after apply)
      + kind                         = "Linux"
      + location                     = "northcentralus"
      + maximum_elastic_worker_count = (known after apply)
      + maximum_number_of_workers    = (known after apply)
      + name                         = "north-asp"
      + reserved                     = true
      + resource_group_name          = "rontest-bold-resources"

      + sku {
          + capacity = (known after apply)
          + size     = "S1"
          + tier     = "Standard"
        }
    }

  # module.appservice.azurerm_app_service_plan.main["southcentralus"] will be created
  + resource "azurerm_app_service_plan" "main" {
      + id                           = (known after apply)
      + kind                         = "Linux"
      + location                     = "southcentralus"
      + maximum_elastic_worker_count = (known after apply)
      + maximum_number_of_workers    = (known after apply)
      + name                         = "south-asp"
      + reserved                     = true
      + resource_group_name          = "rontest-bold-resources"

      + sku {
          + capacity = (known after apply)
          + size     = "S1"
          + tier     = "Standard"
        }
    }

  # module.appservice.azurerm_resource_group.main will be created
  + resource "azurerm_resource_group" "main" {
      + id       = (known after apply)
      + location = "northcentralus"
      + name     = "rontest-bold-resources"
    }

Plan: 5 to add, 0 to change, 0 to destroy.

------------------------------------------------------------------------

Note: You didn't specify an "-out" parameter to save this plan, so Terraform
can't guarantee that exactly these actions will be performed if
"terraform apply" is subsequently run.
```

I also tried the same with terraform 0.13.0 beta 1 by running:
```
terraform-v0.13.0-beta1 init
terraform-v0.13.0-beta1 plan --var-file=multiRegion.tfvars
```

You included a very clear test case, so I'm surprised that I'm not seeing the same behavior as you reported. Do you mind taking a look at my test case and trying it again? I do need a reproduction case on my side that I can give to an engineer in order to work on this, and I really appreciate the work you've done to report this. It's rare that we get such a comprehensive positive and negative test case. I'm just puzzled at why it's not failing for me!

**Comment by mildwonkey** [comment]
I am going to close this issue due to inactivity.

If there is still a question, I recommend the [the community forum](https://discuss.hashicorp.com/c/terraform-core), where there are far more people available to help. If there is a bug or you would like to make a feature request, please open a new issue and fill out the template.
Thanks!
